### ISSUE TITLE
[Bug] [ROCm] DeepSeek-V4.1 FP4 indexer still exhausts HBM under long-context AgentX after #37660

### ISSUE THREAD

**Issue by jhinpan** [sgl-project/sglang#39441]
## Checklist

- [x] I searched related issues and found no report for this remaining shape.
- [x] This uses the only published AMD DeepSeek-V4.1 preview image named by the SGLang cookbook.
- [x] Environment output and a deterministic public-dataset reproducer are included.

## Describe the bug

`DeepSeek-V4.1-Flash` on 4×MI355X (TP4/EP4) still drains driver-visible HBM and aborts in the HIP FP4 indexer under long-context AgentX, even though the image contains #37660 (`[AMD] Fix FP4 indexer OOR`).

The failure is not an initialization problem: the server becomes healthy, serves a normal chat request, and can process hundreds of AgentX requests before a scheduler rank aborts.

The runnable image is:

```text
lmsysorg/sglang:dev-dsv41-mi35x
manifest sha256:c609f74f01b206af9664fb66da8ebef47642a7328a2478eabeebb352e014661a
sglang 0.5.19.dev20260908+g554f817948
```

Important source-provenance correction: the package version is not the runtime source identity. This image overlays `python/sglang` after the base build. The actual files used in the reproduction are:

```text
fp4_indexer_hip.py      sha256:afaf39ecebc62bb8a9c8300e0acf225ff3d0f09053700ef6a15be1cb18080fe8
low_ratio_backend_hip.py sha256:f859185ad42ca6de6e988f156cab0e5c0fd8a49eb7de144a5062c7f5a0162274
deepseek_v4_backend.py   sha256:adb93e5eee51c86c328b2fa9962a12ab8f5f8169cf81de2626fa986ad5299737
```

The base commit does contain #37660, and the overlay retains its fixed-size `_LOGITS_POOL`, `SGLANG_DSV4_FP4_LOGITS_BUDGET_MB=2048`, and `logits_rows_per_chunk`. However, the V4.1 low-ratio HIP caller bypasses that bound:

```text
low_ratio_backend_hip.py::_request_groups
  -> groups only between requests
  -> explicitly keeps one oversized request intact
  -> aiter_fp4_paged_mqa_logits
  -> _alloc_logits: n > _LOGITS_BUDGET_ELEMS
  -> torch.empty(num_tokens, max_seq_len)
```

The observed 14.39 GiB allocation is exactly the resulting rectangle:

```text
16,260 rows × 237,568 compressed positions × 4 bytes = 14.39026 GiB
```

With the configured 2 GiB policy, this request should have been split into eight bounded chunks, not one dynamic rectangle.

Observed in the deterministic reproduction:

- 4×MI355X, TP4/EP4, 8,192-token prefill chunks, concurrency 4 AgentX session trees.
- At failure SGLang reported 7 running requests and 3 queued requests.
- External 0.5-second `rocm-smi` samples rose from about 197.0 GiB used per GPU to 286.0 GiB by 86 seconds.
- GPU 3 peaked at 287.56 GiB used / 0.43 GiB free.
- At 208 seconds the runtime reported `HSA_STATUS_ERROR_OUT_OF_RESOURCES ... Available Free mem : 0 MB`.
- The active stack was `fp4_indexer_hip.py::_alloc_logits -> aiter_fp4_paged_mqa_logits`.
- `scheduler_0` exited `-6`; the parent then sent SIGQUIT to the complete TP service.

This reproduced three ways:

1. Cookbook memory settings (`mem-fraction-static=0.8`, 16K chunks): HSA OOR during warmup.
2. `mem-fraction-static=0.65`, 16K chunks: 257 valid profiled requests over 2,060 seconds at 128.55 output tok/s, then a 14.39 GiB allocation failed with 8.6 GiB free and 37.9 GiB reserved-but-unallocated.
3. `mem-fraction-static=0.65`, 8K chunks, expandable allocator: HSA OOR in `_alloc_logits` during warmup.

One bounded workaround completed a 1,200-second diagnostic (121/121 requests, zero errors), but it changes four capacity controls and is not a fix: `mem-fraction-static=0.60`, 4K chunks, `SGLANG_DSV4_FP4_LOGITS_BUDGET_MB=512`, and `max-running-requests=8`.

Two single-variable controls narrow the diagnosis without closing it:

- `SGLANG_DSV41_TORCH_PREFILL_INDEXER=1` removes the optimized low-ratio caller and delays failure from about 208 to 330 seconds, but HBM still reaches 287.64 GiB and OOR moves to `_low_ratio_index_topk_torch`. Therefore the request-group bypass is real but not the only source of peak pressure.
- `HSA_NO_SCRATCH_RECLAIM=0` prevents the immediate HSA abort for 600 seconds, but warmup stalls at 15/44 requests from 210 seconds onward while GPU3 reaches 287.82 GiB used / 0.17 GiB free. Persistent queue scratch changes the failure shape; it does not remove the underlying HBM exhaustion.

The active AMD V4.1 PR #39186 already contains an additional oversized-request split at commit `150d2ffacf`: one request wider than `rows_per_chunk` is scored in row chunks, candidate publications are concatenated, and consumer candidates are sliced by request-local row offset. That code is not in the tested preview overlay. Its current tests use much smaller requests and do not cover the observed `Q=16,260 / W=237,568` boundary, so #39186 needs an exact image-level runtime recheck rather than being credited from source shape.

#38086 proposes a runner-owned, pool-accounted adaptive workspace; #38101 plus ROCm/aiter#5282 proposes streaming TopK that avoids materializing the complete logits matrix. Both are still open and neither is a proven after for this V4.1 workload.

### HBM time order and single-variable controls

![DeepSeek-V4.1 SGLang MI355X FP4 indexer HBM timeline](https://gist.githubusercontent.com/jhinpan/04320d440cbac7fa959439f492172167/raw/4d5c196153b2a10b4639011ee09290282dda3c8c/dsv41-indexer-oor-evidence.svg)

The top panel is the actual optimized-path failure. The bottom panel keeps the model, requests, seed, TP/EP, chunk size, memory fraction, and 2 GiB policy fixed while changing only the named control.

### Kernel time order and source-backed GPU share

![DeepSeek-V4.1 SGLang MI355X kernel timeline](https://gist.githubusercontent.com/jhinpan/04320d440cbac7fa959439f492172167/raw/84067f52500911e98d2745b081361fac511fd067/dsv41-sglang-kernel-causal.svg)

The left panel is chronological, not an aggregate: FP4 logits/top-k feed sparse attention, then dense GEMM, NCCL, mHC, and MoE repeat across the layer cadence. The right panel is cumulative captured GPU time. The trace is for ownership mapping only; stack/shape collection perturbs wall time.

Public figure sources and provenance: https://gist.github.com/jhinpan/04320d440cbac7fa959439f492172167

## Profile evidence

A separate CPU/GPU torch-profiler capture on the same image and workload mapped the GPU work before terminal failure:

- NCCL collectives: 20.8% of captured GPU time.
- Sparse decode attention: 14.1%.
- Candidate/indexer memory operations: device memset 11.6%, D2D copies 7.4%, and `bitwise_not` 7.3%, mapped to `dsv4/indexer.py::select_candidate_blocks` and `low_ratio_backend_hip.py::_select_topk_extend_hip`.
- mHC boundary: 6.2%.
- FP4 prefill logits and index top-k: 1.9% and 1.6%.

The trace contains an annotated `EXTEND` step with 8,192 query tokens. A representative 30 ms GPU window shows the ordered chain:

```text
FP4 logits -> index top-k -> sparse attention -> dense GEMM
-> NCCL all-reduce -> mHC boundary -> MoE up/down -> repeat
```

The source-backed kernel table matters here: a generic triage heuristic labeled the candidate-selection memset/copies as a grouped-MoE fusion opportunity, but the Python locations place them in the DSV4 indexer path, not the MoE gate.

Do not use the profiled step wall time as a performance number. `with_stack=true` and `record_shapes=true` produced 45 million trace events and heavily perturbed execution. Use it only for kernel-to-source mapping. A formal before/after timing trace should use CPU+GPU activities with stack/shape recording disabled; memory should be sampled externally.

## Reproduction

Server:

```bash
IMAGE='lmsysorg/sglang@sha256:c609f74f01b206af9664fb66da8ebef47642a7328a2478eabeebb352e014661a'
MODEL_DIR=/path/to/DeepSeek-V4.1-Flash

docker run --rm --network host --ipc host --shm-size 64g \
  --device /dev/kfd --device /dev/dri \
  --group-add video --group-add render \
  --cap-add SYS_PTRACE --security-opt seccomp=unconfined \
  -e HIP_VISIBLE_DEVICES=0,1,2,3 \
  -e PYTORCH_HIP_ALLOC_CONF=expandable_segments:True \
  -e SGLANG_DSV4_FP4_LOGITS_BUDGET_MB=2048 \
  -e SGLANG_USE_AITER=1 \
  -e SGLANG_MOE_PADDING=1 \
  -e AITER_FLYDSL_FORCE_REDUCE=1 \
  -e ROCM_QUICK_REDUCE_QUANTIZATION=NONE \
  -e SGLANG_DSV41_REASONING_EFFORT=high \
  -e SGLANG_SIMULATE_ACC_LEN=3.51 \
  -e SGLANG_SIMULATE_ACC_METHOD=match-expected \
  -e SGLANG_SIMULATE_ACC_TOKEN_MODE=real-draft-token \
  -v "$MODEL_DIR:/model:ro" \
  --entrypoint '' "$IMAGE" \
  python3 -m sglang.launch_server \
    --model-path /model \
    --served-model-name deepseek-ai/DeepSeek-V4.1-Flash \
    --host 0.0.0.0 --port 8891 --trust-remote-code \
    --tp 4 --ep-size 4 \
    --disable-radix-cache \
    --mem-fraction-static 0.65 \
    --speculative-algorithm DSPARK \
    --speculative-dspark-block-size 5 \
    --cuda-graph-max-bs 64 \
    --max-running-requests 128 \
    --chunked-prefill-size 8192 \
    --reasoning-parser auto \
    --tool-call-parser auto \
    --watchdog-timeout 3600 \
    --enable-metrics --enable-cache-report
```

Client dependencies:

```bash
git clone https://github.com/SemiAnalysisAI/InferenceX.git
git -C InferenceX checkout 57bec457c08e5223d7143f1e533ae7abf046ebd5
git -C InferenceX submodule update --init utils/aiperf
# AIPerf submodule: 754356e9a39acc6cc6afb242d123bb57c3fb6f75
```

After installing that AIPerf checkout, run:

```bash
aiperf profile \
  --scenario inferencex-agentx-mvp \
  --url http://localhost:8891 \
  --endpoint /v1/chat/completions \
  --endpoint-type chat --streaming \
  --model deepseek-ai/DeepSeek-V4.1-Flash \
  --tokenizer deepseek-ai/DeepSeek-V4.1-Flash \
  --concurrency 4 --benchmark-duration 3600 \
  --stats-interval 30 --random-seed 42 \
  --failed-request-threshold 0.10 \
  --trajectory-start-min-ratio 0.25 \
  --trajectory-start-max-ratio 0.75 \
  --warmup-requests-per-lane 10 \
  --trace-idle-gap-cap-seconds 300 \
  --warmup-grace-period 1800 \
  --use-server-token-count --no-gpu-telemetry \
  --tokenizer-trust-remote-code \
  --num-dataset-entries 393 --slice-duration 1.0 \
  --server-metrics http://localhost:8891/metrics \
  --output-artifact-dir ./aiperf_artifacts \
  --public-dataset semianalysis_cc_traces_weka_062126
```

## Environment

```text
Python: 3.12.3
ROCM available: True
GPU 0,1,2,3: AMD Instinct MI355X
ROCM_HOME: /opt/rocm-7.2.4
HIPCC: HIP 7.2.53211-97f5574fe2
ROCM Driver Version: 6.16.13
PyTorch: 2.11.0+rocm7.2
sglang: 0.5.19.dev20260908+g554f817948
sglang-kernel: 0.4.6.post1
triton: 3.7.0+amd.rocm7.2.0.git89002410
transformers: 5.12.1
```

## Expected behavior / acceptance test for a fix

Use the command above unchanged before and after the candidate patch.

1. The 3,600-second profile completes with zero server errors.
2. Driver-visible HBM reaches a bounded plateau after scratch initialization instead of climbing to total capacity.
3. A low-overhead formal trace (`activities=["CPU","GPU"]`, no stack/shape recording) shows the intended allocation or chunking mechanism changed; a separate one-step mapping trace may retain stacks.
4. DSpark acceptance length and model output remain within the same validated band.
5. Report both throughput and the memory/timeline mechanism. A throughput delta alone does not prove the motivation was fixed.

Before the full run, exercise the allocation boundaries at fixed compressed width 237,568 with query rows `564, 565, 2259, 2260, 4096, 8192, 16260`. The 2 GiB policy must never issue an over-budget logits allocation; the 16,260-row request should produce eight score/top-k chunks. Run each shape cold and then as one increasing-context sequence to separate legitimate active pressure from allocator fragmentation.

**Comment by jhinpan** [comment]
Retest update: the reported FP4-indexer failure is fixed at the exact
current #39186 head, `92fb91b457fdf3e90e48e5f4f90bdd0ba46ff1f7`.

![PR #39186 HBM retest](https://gist.githubusercontent.com/jhinpan/04320d440cbac7fa959439f492172167/raw/e9b885a92d84d547bbad6a36145d09a04b89ce6d/dsv41-pr39186-retest-hbm.svg)

I built `docker/rocm.Dockerfile` with local source at that commit and
`GPU_ARCH=gfx950-rocm724`. The resulting image reports SGLang
`0.5.20.dev271+g92fb91b45`, PyTorch `2.11.0+rocm7.2`, HIP `7.2.26015`,
Triton `3.7.0`, and AITER `4ad9983282` plus the Dockerfile-declared
patches. The loaded FP4 indexer, low-ratio caller, V4 backend, and tuned
FMoE CSV hashes match the PR checkout.

Canonical AgentX result, with the same model, 393-trace corpus, seed 42,
TP4/EP4, concurrency 4, DSPARK settings, disabled radix cache,
`--chunked-prefill-size 8192`, and `--mem-fraction-static 0.65`:

- 3,601.067 seconds
- 380/380 profiling requests and 44/44 warmup requests
- zero request errors and zero matched server errors
- server alive before controlled teardown, Docker restart count 0
- 129.04118 output token/s
- TTFT p50 3.39362 s; ITL p50 3.88 ms
- DSpark accept length mean 3.5112 (server-log samples), metric p50 3.50
- HBM peak 206.97–207.22 GiB/GPU; minimum free 80.76 GiB/GPU
- final 600-second HBM slope +0.000281 GiB/min; range 0.0041 GiB

The causal control used the same built image and changed only one
boundary: it replaced #39186's 30-line within-request
split/concatenate/slice branch with the former 8-line one-shot call.
That control failed on the fourth warmup request: HBM reached
284.29–284.54 GiB/GPU, all four workers reported `Memory access fault`,
the schedulers aborted with `SIGABRT`, and two requests returned HTTP
500. This isolates the stability change to the within-request bound
rather than to a different image or allocator setting.

For the reported `W=237,568` width, the fixed runtime emitted only
2,259-row score chunks plus the remainder; no score call exceeded the
2 GiB policy. A stacked regression test also drives the original
`Q=16,260` boundary and requires
`[2259, 2259, 2259, 2259, 2259, 2259, 2259, 447]`; removing the split
restores one 16,260-row / 14.390259-GiB call. The test is available in
[`7791819c10`](https://github.com/jhinpan/sglang-fork/commit/7791819c109f34a1d04c52d71b1b8acc9e741b97).

The pre-fix partial canonical result was 128.55 output token/s, so the
completed result is only +0.38%. This is a memory/stability fix, not a
throughput claim. The vLLM gap also remains.

Full result and provenance:
https://gist.github.com/jhinpan/04320d440cbac7fa959439f492172167

I am leaving this issue open for now because #39186 is still open and
currently reports a dirty merge state, and the affected preview image
remains unchanged. Once the fix is merged into the served branch or a
replacement image is published, this issue can be closed without
implying that the currently published runtime is safe.
